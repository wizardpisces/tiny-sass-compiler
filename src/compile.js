import parse from './parse'
import transform from './transform'
import {
    NodeTypes
} from './parse/ast';
/**
 * 
 * ast => css
 * 
 */
function compile_css(ast) {

    function compile(exp) {
        switch (exp.type) {
            case NodeTypes.TEXT:
                return css_str(exp);
            case NodeTypes.VARIABLE:
                return css_var(exp);
            case NodeTypes.ASSIGN:
                return css_assign(exp);
            case NodeTypes.CHILD:
                return css_child(exp);
            case NodeTypes.EMPTY:
                return '';

            default:
                throw new Error("Don't know how to compile expression for " + JSON.stringify(exp));
        }
    }

    function css_str(exp) {
        return exp.value;
    }

    function css_var(exp) {
        return exp.value;
    }

    function css_assign(exp) {
        return compile(exp.left) + ':' + compile(exp.right) + ';';
    }

    function css_child(exp) {
        return exp.selector.value + '{' + exp.children.map(child => compile(child)).join('') + '}';
    }

    function toplevel(ast) {
        return ast.prog.map(exp => compile(exp)).join('');
    }

    return toplevel(ast);
}

export default (scss, sourceDir) => compile_css( transform( parse(scss), sourceDir ) )