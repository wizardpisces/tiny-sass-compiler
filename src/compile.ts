import { CompilerOptions } from './options'
import baseParse from './parse'
import { generate } from './codegen'
import { transform } from './transform'
import {Environment} from './parse/util'

import {transformStatement} from './tranform-plugin/transformStatement'

export default function baseCompile(scss: string, options: CompilerOptions = { env : new Environment(null)}) {
    
    let ast = baseParse(scss)
    
    ast.source = scss;

    transform(ast, {
        nodeTransforms: [transformStatement],
        ...options
    })

    return generate(ast,options)
}