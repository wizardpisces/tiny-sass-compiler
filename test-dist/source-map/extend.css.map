{
  "version": 3,
  "sources": [
    "test/extend.scss"
  ],
  "names": [],
  "mappings": "AAAA;EACI,OAAQ;EACR,QAAS;EACT,MAAO;;AAGX;EACE,OAAQ;EACR,QAAS;EACT,MAAO;;AAOT;EAEI,aAAc;;AAElB;EACI,MAAM",
  "sourcesContent": [
    ".message-shared {\n    border: 1px solid #ccc;\n    padding: 10px;\n    color: #333;\n}\n\n%message-shared {\n  border: 1px solid #ccc;\n  padding: 10px;\n  color: #333;\n}\n\n.message {\n    @extend .message-shared;\n}\n\n.success {\n    @extend .message-shared;\n    border-color: green;\n}\n.error{\n    color:red;\n    @extend %message-shared\n}"
  ]
}