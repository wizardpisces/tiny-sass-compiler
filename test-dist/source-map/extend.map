{
  "version": 3,
  "sources": [
    "template.scss"
  ],
  "names": [],
  "mappings": "AACG,AAAQ,AACR,AAAS,AACT,AAAO,AAIT,AAAQ,AACR,AAAS,AACT,AAAO,AASL,AAAc,AAGd,AAAM",
  "sourcesContent": [
    ".message-shared {\n    border: 1px solid #ccc;\n    padding: 10px;\n    color: #333;\n}\n\n%message-shared {\n  border: 1px solid #ccc;\n  padding: 10px;\n  color: #333;\n}\n\n.message {\n    @extend .message-shared;\n}\n\n.success {\n    @extend .message-shared;\n    border-color: green;\n}\n.error{\n    color:red;\n    @extend %message-shared\n}"
  ]
}