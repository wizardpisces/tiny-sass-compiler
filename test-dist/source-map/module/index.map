{
  "version": 3,
  "sources": [
    "template.scss"
  ],
  "names": [],
  "mappings": "AAKC,AAAM,AACN,AAAO,AAFP,AAAkB,AAClB,AAAO",
  "sourcesContent": [
    "// styles.scss\n@import \"./base.scss\";\n\n.inverse {\n  background-color: $primary-color;\n  color: white;\n}"
  ]
}