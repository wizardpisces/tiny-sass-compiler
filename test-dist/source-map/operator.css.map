{
  "version": 3,
  "sources": [
    "operator.scss"
  ],
  "names": [],
  "mappings": ";EAEE,MAAO;;;EAIP,MAAO;EACP,MAAO;;;EAIP,MAAO;EACP,MAAO",
  "sourcesContent": [
    "$var : 600px;\n.container {\n  width: 1px + 8px;\n}\n\narticle[role=\"main\"] {\n  float: left;\n  width: $var / 960px * 100%;\n}\n\naside[role=\"complementary\"] {\n  float: right;\n  width: 300px / 960px * 100%;\n}"
  ]
}