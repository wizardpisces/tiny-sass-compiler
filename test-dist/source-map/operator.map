{
  "version": 3,
  "sources": [
    "operator.scss"
  ],
  "names": [],
  "mappings": "AACA;EACE,MAAO;;AAGT;EACE,MAAW;EACX,MAAO;;AAGT;EACE,MAAY;EACZ,MAAO",
  "sourcesContent": [
    "$var : 600px;\n.container {\n  width: 1px + 8px;\n}\n\narticle[role=\"main\"] {\n  float: left;\n  width: $var / 960px * 100%;\n}\n\naside[role=\"complementary\"] {\n  float: right;\n  width: 300px / 960px * 100%;\n}"
  ]
}