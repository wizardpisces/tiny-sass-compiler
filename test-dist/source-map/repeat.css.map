{
  "version": 3,
  "sources": [
    "repeat.scss"
  ],
  "names": [],
  "mappings": "AAAA;EACI,MAAO;;AAGX;EACI,MAAO;EACP,KAAM",
  "sourcesContent": [
    ".repeat1 {\n    color: red;\n}\n\n.repeat1 {\n    color: green;\n    size: 14px;\n}\n"
  ]
}