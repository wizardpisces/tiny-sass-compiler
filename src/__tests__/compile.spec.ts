import compile from '../index'
import { SourceMapConsumer, RawSourceMap } from 'source-map'

describe('compiler: integration tests', () => {
    const source = `
$font-stack:    Helvetica, sans-serif;
$primary-color: #333;

body .test{
  font: 100% $font-stack;
  color: 
  $primary-color;
}`.trim()
    interface Pos {
        line: number
        column: number
        name?: string
    }

    function getPositionInCode(
        code: string,
        token: string,
        expectName: string | boolean = false
    ): Pos {
        const generatedOffset = code.indexOf(token)
        let line = 1
        let lastNewLinePos = -1
        for (let i = 0; i < generatedOffset; i++) {
            if (code.charCodeAt(i) === 10 /* newline char code */) {
                line++
                lastNewLinePos = i
            }
        }
        const res: Pos = {
            line,
            column:
                lastNewLinePos === -1
                    ? generatedOffset
                    : generatedOffset - lastNewLinePos - 1
        }
        if (expectName) {
            res.name = typeof expectName === 'string' ? expectName : token
        }
        return res
    }
    test('function mode', () => {
        const { code, map } = compile(source, {
            sourceMap: true,
            filename: `foo.scss`
        })

        expect(code).toMatchSnapshot()

        expect(map!.sources).toEqual([`foo.scss`])
        expect(map!.sourcesContent).toEqual([source])

        const consumer = new SourceMapConsumer(map as RawSourceMap)
        
        console.log(getPositionInCode(code, `test`))
        console.log(consumer.originalPositionFor(getPositionInCode(code, `test`)))
        console.log(getPositionInCode(source, `test`))
        expect(
            consumer.originalPositionFor(getPositionInCode(code, `test`))
        ).toMatchObject(getPositionInCode(source, `test`))
    })
})