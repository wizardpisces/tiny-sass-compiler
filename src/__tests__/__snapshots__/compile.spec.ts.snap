// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: integration tests function mode 1`] = `"body .test{font:100% Helvetica, sans-serif;color:#333;}"`;
